### ISSUE TITLE
[1.7.10] Concurrent Modification Exception being thrown in Explosion Source Manager

### ISSUE THREAD

**Issue by UndeadZeratul** [elix-x/Avoid-Exploding-Creepers#1]
From [Here](https://github.com/UndeadZeratul/ThereWillBeBlood/issues/128)

I have Creeper Awareness v2.0 for 1.7.10, but it looks like you've recently pushed a change to the namespace and API; although I don't think it will matter much, because I don't think the offending method changed during that commit.

It looks like [this](https://github.com/elix-x/Creeper-Awareness/blob/master/1.7.10/src/main/java/code/elix_x/mods/creeperawareness/api/ExplosionSrcManager.java#L128-L148) is the method that's broken, and it's probably because you're looping through a list that is being modified within the loop.  I'd say grab an iterator and instead of using a for loop, do a `while(iterator.hasNext())` structure, that way if an entry in the list is altered or removed, the iterator won't care.

**Comment by Elix-x** [comment]
Could you wait for official release?
Curse page is ready and i'm nearly done fixing forge corruption caused by windows restore.
It will be out this saturday for sure.

Because i changed quite a bit of code, so that could have fixed it.

**Comment by UndeadZeratul** [comment]
Oh, yeah that's no problem.  This happens on the current release of my modpack, but I'm still working on the next version, so there's no immediate rush.

**Comment by Elix-x** [comment]
Ok.

**Comment by Elix-x** [comment]
Creeper Awareness was released... 2 weeks ago... And i forgot about this...
Could you check again please?

**Comment by UndeadZeratul** [comment]
Oh, right; sorry I've been busy in real life.  I can run some tests with the new version later and I'll let you know if I find anything.

**Comment by Elix-x** [comment]
No problems!
Thanks!

**Comment by UndeadZeratul** [comment]
Hey, sorry it took so long to get back.

So I grabbed your "Creeper Awareness 2.0-1.7.10-CORRECT" file that's on curse, [this one](http://minecraft.curseforge.com/projects/creeper-awareness/files/2291603), and as I was flying around testing some stuff out, I got a [crash log](http://pastebin.com/L3EN2KQX) that still seems to have the same exception.

**Comment by Elix-x** [comment]
Thanks!
I'll ask one thing, though: do you have any mods that put world executions on separate threads?

**Comment by UndeadZeratul** [comment]
I'm not entirely sure, do you know of any examples?

Actually, I can just give you the whole list [here](https://github.com/UndeadZeratul/ThereWillBeBlood/blob/master/Versions.md), if that helps instead.

**Comment by Elix-x** [comment]
It seems that you don't...

Example would be some kind of alternative server environment. (I'm not sure if, but rumors tell that spygot and sponges are one of those).

Alright, i'll look into that.

**Comment by UndeadZeratul** [comment]
Ah, no I don't use anything but pure Minecraft Forge (also this was in Single Player when I found it, I haven't tried on a server yet, which would still be pure Forge).

**Comment by Elix-x** [comment]
Ok. That's even weirder than.
You're on 1.7.10, right?

**Comment by UndeadZeratul** [comment]
That is correct, I am.

**Comment by Elix-x** [comment]
Alright, i'll see if i can find it on my end. Maybe i can directly fix it.

**Comment by UndeadZeratul** [comment]
Just from looking at that method, it looks like it's looping through the world's list of entities, which possibly changes sometime during the processing, which causes the concurrent modification exception.  My guess is a mob despawned or was killed or something.

I've had pretty decent luck by instead grabbing an iterator off of the list, and changing the for loop to a while loop, checking if the iterator hasNext().  Sadly, I'm not entirely sure how to re-create this one, I literally just flew around a new world to test out some terrain generation and crashed with that exception.

**Comment by Elix-x** [comment]
Another problem is that i've never met enything like that. For some of videos that didn't make it to  curse/mcf page, i was exploding (with commands) hundreeds of creepers and tnt, in different dimensions and on different version. And i have never met this crash.

**Comment by UndeadZeratul** [comment]
Yeah, it's very strange, and like I said, hard to reproduce.  I'm lucky I even got it to crash with it.  My original issue was reported via a YAMPST crash report, so someone playing my pack experienced it and didn't report it to me, I just happened to see the crash report and figured I'd forward it onto you.

It may not be the entities close to the explosion source that are causing it, it might be just some random entity far away from the player despawning, who knows.  I just know what concurrent modification exceptions are because I've dealt with them at work, although those were more because I was modifying the array while I was looping through it, deleting entries that I'd soon have to iterate to, and thus it'd fail (whoops).  With something as random and dynamic as this, it's definitely harder to pin down WHY it's happening.

EDIT: Luckily I can immediately boot the pack back up and log back into the world and not immediately crash again, so that's good.

**Comment by Elix-x** [comment]
At least that's good :D.

**Comment by UndeadZeratul** [comment]
Any chance of this being backported to 1.7.10?  If not I'll just close the issue on my repo as a can't/won't fix.

**Comment by Elix-x** [comment]
So this happens in 1.7.10 too? Ok. I'll backport it, but probably tomorrow (i planned to fix it in all versions, but didn't have time today, so should bacport it tomorrow).

**Comment by UndeadZeratul** [comment]
Oh, yeah I thought I mentioned that; my modpack is 1.7.10.  Sorry about that!  And no rush, I'm patient.

**Comment by Elix-x** [comment]
No, that's me being dumb. Title says it all :laughing: .

**Comment by UndeadZeratul** [comment]
Oh, haha; then I'm blind too because I didn't even see that either.

**Comment by UndeadZeratul** [comment]
Well..  I just got a crash report while using 2.0.1 and it's in the same spot...  I don't know what could be causing it nor do I know what else to try to fix it, I just figured you should know...

[New Crash Report](https://github.com/UndeadZeratul/ThereWillBeBlood/issues/145)
[Reported Line of Code](https://github.com/elix-x/Creeper-Awareness/blob/master/1.7.10/src/main/java/code/elix_x/mods/creeperawareness/api/ExplosionSrcManager.java#L131)

**Comment by Elix-x** [comment]
Iterator didn't fix this issue, although he didn't crash nor blocked on hasNext. I think i'll reopen and mark this as wont fix.

**Comment by UndeadZeratul** [comment]
That's fine; I don't know why it happened again since I would've thought that would fix it.  Since you could barely recreate the issue yourself beforehand and I've only had it happen once myself, it's a pretty rare crash; one that doesn't seem to leave any permanent damage either.  I just figured I'd let you know of the issue since it did pop up again, as odd as that may be.

**Comment by Elix-x** [comment]
...as odds that may be... too low :laughing: .
